Uniform Fission Sites

valeriaRaffuzzi opened the pull request:
This pull request includes a few changes. 

First, I am adding UFS as a vector field. This requires the specification of a spatial map over some fissile material. The input information are read inside the eigenPP; additionally, a UFS flag must be specified in the implicit collision operator for this to work. 

This spatial/material map does not have to be homogeneous in terms of fissile material volume; if it isn't, one should specify it in the input with a 'uniformMap 0' flag. In this case, a simple routine to estimate the fissile material volume of each bin in the map is run during initialisation. 

Then, when generating fission neutrons in collOpImp, the map is read and updated. 

Additional features in this pull request are: 
- adding unstruct and equivolume bins in the tally sphericalMap
- generating a 3D cylindrical tally map, i.e. with radial, axial and azimuthal bins

Mikolaj-A-Kowalski commented on CollisionOperator/CollisionProcessors/neutronCEimp_class.f90:
```diff
@@ -485,10 +517,20 @@ subroutine split(self, p, thisCycle)
+    integer(shortInt)                     :: mult, i, n
+    real(defReal)                         :: prob
+
+    n = floor(p % w/self % maxWgt)
+    prob = p % w/self % maxWgt - n
+
+    if (p % pRNG % get() < prob) then
+      mult = n + 1
+    else
+      mult = n
+    end if
+
+    if (mult == 1) return
```
I have a vague recollection that similar change was in the weight windows PR and we decided to stay with the `ceiling` approach. Or am I forgetting something?

valeriaRaffuzzi replied:
No, I think I changed the two branches the same way but I'll go back to normal!

Mikolaj-A-Kowalski commented on CollisionOperator/CollisionProcessors/neutronCEimp_class.f90:
```diff
@@ -168,6 +179,13 @@ subroutine init(self, dict)
 
+    ! Sets up the uniform fission sites field
+    if (self % uniFissSites) then
+      name = 'UFS'
```
Similarly to weight windows, perhaps we should make the common name a constant in `universalVariables`?

Mikolaj-A-Kowalski reviewed (commented):
I think it looks all right. Thanks for the development! 

It might be good idea though to rebase it on current `main` so the GitHub Actions tries to compile and test it before we marge.

valeriaRaffuzzi commented:
I addressed all the comments. 

However, we still need to fix the issue of the circular dependencies. I will do that in the near future, and then merge this and resume the VarReduction pull request.

Mikolaj-A-Kowalski commented on Geometry/fieldFactory_func.f90:
```diff
@@ -0,0 +1,80 @@
+!!
+!! Module to build new fields. Called by geometryReg_mod
+!!
```
I think it is now the other way around, right? 
This module calls `geomReg` and adds the geometry.

Mikolaj-A-Kowalski commented on Geometry/fieldFactory_func.f90:
```diff
@@ -0,0 +1,80 @@
+
+contains
+
+  !!
+  !! Allocates and initialises field from dictionary
+  !!
+  !! Args:
+  !!   dict [in]   -> Dictionary with definition
+  !!   name [in]   -> Name of the field for the geometry registry
+  !!
+  !! Errors:
+  !!   fatalError is type of field is unknown
+  !!
+  subroutine new_field(dict, name)
```
Maybe it is worth to explicitly mention that the field gets allocated inside the `geometryReg`?

Mikolaj-A-Kowalski commented on Geometry/geometryFactory_func.f90:
```diff
@@ -0,0 +1,85 @@
+
+contains
+
+  !!
+  !! Allocates and initialises a geometry from dictionary
+  !!
+  !! This is Factory procedure for geometries
+  !!
+  !! Args:
+  !!   dict [in]  -> Dictionary with geometry definition
+  !!   name [in]  -> Name of the geometry for the geometry registry
+  !!   silent [in] -> Optional. Set to .true. to surpress console messeges. Default .false.
+  !!     Note that errors will still be printed with silent=.true.
+  !!
+  !! Errors:
+  !!   fatalError is type of geometry is unknown
+  !!
```
Same as for fields.

Mikolaj-A-Kowalski commented on Geometry/geometryFactory_func.f90:
```diff
@@ -0,0 +1,85 @@
+!!
+!! Module to build new fields. Called by geometryReg_mod
+!!
```
Same as for fields.

Mikolaj-A-Kowalski commented on Geometry/geometryReg_mod.f90:
```diff
@@ -106,23 +96,13 @@ module geometryReg_mod
-  subroutine addGeom(name, dict, silent)
-    character(nameLen), intent(in)         :: name
-    class(dictionary), intent(in)          :: dict
-    logical(defBool), optional, intent(in) :: silent
-    integer(shortInt)                      :: idx
-    logical(defBool)                       :: silent_l
+  subroutine addGeom(geom, name)
+    class(geometry), allocatable, intent(inout) :: geom
+    character(nameLen), intent(in)              :: name
+    integer(shortInt)                           :: idx
     integer(shortInt), parameter :: NOT_PRESENT = -7
     character(100), parameter    :: Here = 'addGeom (geometryReg_mod.f90)'
```
The doc comment needs to be updated.

Mikolaj-A-Kowalski commented on Geometry/geometryReg_mod.f90:
```diff
@@ -224,10 +206,10 @@ end function geomNum
   !!   fatalError if field with the name was already defined
   !!
-  subroutine addField(name, dict)
-    character(nameLen), intent(in)         :: name
-    class(dictionary), intent(in)          :: dict
-    integer(shortInt)                      :: idx
+  subroutine addField(kentta, name)
+    class(field), allocatable, intent(inout) :: kentta
+    character(nameLen), intent(in)           :: name
+    integer(shortInt)                        :: idx
     integer(shortInt), parameter :: NOT_PRESENT = -7
     character(100), parameter    :: Here = 'addField (geometryReg_mod.f90)'
```
Again doc comment diverged from the subroutine.

Mikolaj-A-Kowalski reviewed (approved):
I think all it needs is to update the documentation comments and it is safe to be merged. 
Thanks for the update!
